### ISSUE TITLE
[CLOSED] basisproject --noalter option for creating new templates

### ISSUE THREAD

**Issue by schuhschuh** [cmake-basis/BASIS#219]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [schuhschuh](https://github.com/schuhschuh)**
_Saturday Nov 09, 2013 at 18:04 GMT_
_Originally opened as https://github.com/schuhschuh/cmake-basis-legacy/issues/219_

---

This option would be useful to create a new template populated with the current version of selected/all template files.

Example: `basisproject --full --noalter`

This option would disable any substitutions during the population of a project directory. For the creation of a new template from another template, another option would be needed, however, which also copies/creates the required `_config.py` file.

**Reporter:** Andreas Schuh
**Assigned to:** Andreas Schuh
**Begin:** 2013-10-21
**Completed:** 0

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Saturday Nov 09, 2013 at 18:13 GMT_

---

Implemented by d45c761ff27d8abe29d0591d6e683952d26db9f6.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 00:00 GMT_

---

The previous addition of the `--orig` option was a bit hasty. There is a conflict with the `--original` option, which may lead to confusion.

Fix up the previous commit with a new option which helps one to populate a directory with the full set of files which are part of a given template (such as the default one shipped with BASIS) in order to create their own organization specific template.

Note to myself: Don't forget to copy the new `_config.py` file as well, respectively, figure how to merge the separate configuration scripts.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 00:12 GMT_

---

As it turns out I keep forgetting what I've actually done. Good time for making better use of Git's feature branches :S... The `--orig` option has been renamed to `--noalter` as part of commit ded2fcb76fb97cedb5cd3683834de4a501eb16f1.

Remains the issue of dealing with the template configuration file(s).

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 13, 2013 at 00:15 GMT_

---

perhaps --createTemplate would be a clear name for what it does?

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 00:17 GMT_

---

Well, previously (before the extra template configuration file), it actually did nothing else then what it would do when you create a new project. The only difference being whether substitutions are performed or not. As these are done by the function `basisproject.alter`, the name `--noalter` made sense to me.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 00:22 GMT_

---

But your comment made me think of the following. There could be an option to create a template configuration file from the information (dictionaries) loaded from an existing template (hierarchy). Only the substitutions would get lost, as only the actual values would then be present. I guess this needs a bit more thought, but also has lower priority. I think everyone who actually will create a template for others will have no problems to just copy the template configuration file from `data/template-1.0` and modify it. More important to get the feature fully working.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 13, 2013 at 00:22 GMT_

---

Yeah in context that does make sense, but perhaps my suggestion makes sense with how things work now?

Also, don't forget that the way `_config.py` is currently set up will mean a generated template will not work, because the other earlier versions of `_config.py` will not be present. I think we will need to resolve that.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 00:25 GMT_

---

I was thinking about something similar to the following, but properly formatted output to a file instead:

``` python
config = load_config(get_template(args.template, "_config.py"))
print config.required
print config.options
print config.substitutions
```

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 13, 2013 at 00:44 GMT_

---

Wouldn't that lose any comments or other information in the file? I was kind of thinking it may be best to have the full config in each version so all the behavior stays straightforward. While there is more duplicate text it is consistent with how things work for all the other files.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 01:04 GMT_

---

You got a point by "consistent with how things work for all the other files". Indeed even if just one line is modified, you have to copy the whole rest of the file. Using incremental file changes would make things far too complicated. They probably are already anyways for the purpose.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Saturday Nov 16, 2013 at 01:31 GMT_

---

As part of commit c068ac8beacc5859d7255b62b046df77d2c5914d, the template configuration files do no longer load the respective previous version, but a fresh copy is needed.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Saturday Nov 16, 2013 at 19:49 GMT_

---

To clarify, does that mean each template version folder needs every single file now including _config.py?

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Tuesday Nov 19, 2013 at 01:05 GMT_

---

The following line in `basisproject` will look through all versions from latest to oldest until it finds a `_config.py` file.

``` python
config_file = get_template(args.template, '_config.py')
```

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Tuesday Nov 19, 2013 at 01:09 GMT_

---

I am not sure if I understand your question fully. The complete template consists of all files in every version subdirectory of the template directory. If one file is in more than one version subdirectory, the one corresponding to the latest version is used. This applies to all template files including the `_config.py` file. When any template file shall be modified for a new version, the file has to be copied to a new version subdirectory and may then be edited. The same applies to the template configuration file. The only change is that in the template configuration files, I no longer include previous template configurations, but instead copy all the code. Given this as requirement, it is then easy to implement this issue here as only one template configuration file needs to be copied instead of trying to combine multiple of these.

You may have been mislead by me adding a copy of the template configuration to every version subdirectory of the `sbia` template. This was indeed not necessary and I consider removing the obsolete template configuration files again. I realized so only after I added them already.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Thursday Nov 21, 2013 at 22:08 GMT_

---

oh ok. That makes sense. Thanks!

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 27, 2013 at 03:59 GMT_

---

I believe I found a bug during testing. If you supply other parameters, like `--author "author name"` to the template in addition to `--original` it will perform substitution, rather than ignoring all substitution parameters.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 27, 2013 at 05:03 GMT_

---

actually the problem seems to go further than I originally thought, it appears something may have broken templates in basisproject generally after the install step. 

My directory contains:

```
basis
basis-build
basis-install
```

I configured cmake to use those directories.

Here is my response file:

```
--name "nrecPractice"
--description "I'm a desc"
--use FakeThing
--full
--noalter
--template /Users/ahundt/Source_Code/coretech/basis/nrecBASISTemplate/1.0
```

If I open basisproject the `_template_dir` correctly mentions `../share/templates` and `_template` refers to `nrecBASISTemplate/1.0`. However when I run `basisproject -h` the only arguments I get from the help are as follows:

```
optional arguments:
  -h, --help            show this help message and exit
  --responsefile <file>
                        File containing additional command line parameters to
                        parse before execution. Include other response files
                        recursively using @/path/to/file.rsp (default: None)
```

Even some of the hard coded ones like `--template` are missing from the help output, perhaps it is exiting without an error or warning? This is `problem D` in this comment.

This is even true if I specify `./basis-install/bin/basisproject --template basis-install/share/templates/sbia/1.8 -h`, I have also verified that all the templates appear to be present in `basis-install/share/templates/`, both my version and the sbia template.
...
I think I have been able to isolate the issue further. It appears to work if I run basisproject in the build dir, but once basisproject is in the install directory it exhibits these problems, in my case in one folder I had `basis` `basis-build` and `basis-install`. This is `problem A`  in this comment.

Problems A and D are for reference later since this comment is so long. :-)

For reference I am using commit db02365 to test.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 27, 2013 at 05:17 GMT_

---

Also when stuff is changed from the description in the original issue, I suggest updating the original issue. I used `--orig` for some time before I figured out it was `--noalter`.

Suggestion F: change `--noalter` to `--noTemplateAlterations` or `--noAlterations` or `--copyTemplateVerbatim`.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Thursday Nov 28, 2013 at 00:20 GMT_

---

Regarding your initial comment:

> I believe I found a bug during testing. If you supply other parameters, like --author "author name" to the template in addition to --original it will perform substitution, rather than ignoring all substitution parameters.

As I think you noticed now, the option `--original` had a different purpose. I noted this earlier saying that the prematurely added `--orig` option introduced unwanted confusion. Therefore I renamed it to `--noalter`, which I think is reasonable and I prefer it over an unnecessary long option name.

Overall, we can think of a different option for creating a new template based on an existing one. This was I think the idea we agreed on last. The thing is simply that this issue here is yet open and has not been finalized.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Thursday Nov 28, 2013 at 00:22 GMT_

---

Regarding problem D: I am aware that the help will not be complete. The problem is caused by the response file option, which requires an initial parsing of the options. Unfortunately, the `-h` option is already parsed at this point though it should yet be ignored. There is certainly a way to prevent argparse from handling the help options but I had not attempted to fix this before.

I fixed this issue now by commit 510bddef5a1700e4a3d478d28f86e7af0e75c39c.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Monday Dec 02, 2013 at 21:25 GMT_

---

Ok I've been able to verify that it is working, switching this back to the documentation step.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Tuesday Dec 03, 2013 at 21:11 GMT_

---

Commit 69a0323 includes an example of the use of `--noalter`. Do we need more documentation of it?

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Dec 04, 2013 at 20:49 GMT_

---

It appears that `_config.py` is not copied when `--noalter` is run.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Dec 04, 2013 at 22:47 GMT_

---

Indeed it does not, but this is also intended. Since the first addition of this option (first called `--orig`) a few things have changed. I will keep the `--noalter` option, but not for the purpose we initially had in mind. Therefore, as stated previously, we will add a new option that can be used for creating a new template given an existing one.

What about actually calling that other option "copy" as creating the new template indeed just means copying the existing template ?

`basisproject --name other --template default/1.4 --copy`
or
`basisproject --name other --template default/1.4 --copy-template``
or
``basisproject --template default/1.4 --new-template other``

which will copy the files of template `default/1.4` to a subdirectory named `other/1.0` including the `_config.py` file. The last would also allow you to specify `other/1.0` or `other/1.1`.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Dec 04, 2013 at 22:50 GMT_

---

Actually, I am wondering if we couldn't use the existing options. Let's say you call the command as follows:
`basisproject --original default/1.4 --template other/1.0`
This will only create a new template when no project name was specified. Otherwise, the `--template` option is supposed to be an existing template already. The `--original` option is optional. If not given, the default template is used.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Dec 04, 2013 at 22:57 GMT_

---

What would `--noalter` do then that is different from `--copy-template`?

I think `--copy-template` would be better since it is more explicit. Implicit behavior is more confusing.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Dec 04, 2013 at 23:19 GMT_

---

`--noalter` does just what it did before and what it says: create a new project, but not alter any of the template files. Maybe a user prefers to edit the files manually ? That would be pretty much the only application of it.

I think I prefer `--new-template` for the last reason I mentioned. We can also call it `--create-template` if you prefer and as you suggested initially.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Dec 04, 2013 at 23:37 GMT_

---

I like `--create-template`. In addition to creating the version number I think `--create-template` should default to work as if `--full` was passed to `--noalter`, so it copies the whole template and doesn't leave any files out by default.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Dec 04, 2013 at 23:42 GMT_

---

I think we should leave it as-is, and create an issue describing `--create-template` marked as a 3.1 feature. I think the current functionality is adequate and documented, even if it isn't perfect, and we will be that much closer to release. :-)

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Thursday Dec 05, 2013 at 00:13 GMT_

---

I created the `--new-template` issue #259 since it is a separate task. At the moment I assigned it for 3.1. 

I also think `--noalter` should still copy `_config.py`, because it is trivial to delete but annoying to copy manually.
